rfcs/0017-DML.md

```diff
@@ -0,0 +1,55 @@
+---
+feature: DML design 
+authors:
+  - "st1page"
+start_date: "2022/10/31"
+---
+
+# DML design
+
+## Motivation
+
+the [Unify the materialized source and table](https://github.com/risingwavelabs/rfcs/pull/4) will lead to 2 things
+
+- table's data are materialized in storage and we should check the consistency and constraints when materializing
+- table can both subscribe to some connectors and support `INSERT`, `DELETE,` and `UPDATE`, which meas it should serve two sources of data changes.
+
+This rfc is to give the details of that.
+
+## Design
+
+### Different objects's DML
+
+#### Source
+
+A source get append-only records from external source and `INSERT` statment and our system will generate a row_id for each record. User can not define PK and other constraints on that. Because all row id are generated internally, we do not need to check the pk's unique constraint.
+
+#### Table
+
+A table can get any operation from external source and `INSERT`, `UPDATE`, `DELETE` statement. The later operations can overwrite the previous ones. If user does not define a primary key, we will generate a row_id for each record. But because user can do modification on table, we must always do the check no matter if the pk is generated internally.
```

Comment:
A little unclear here. Insert into table is also insert into Source?

Reply:
After we adopt the design https://github.com/risingwavelabs/rfcs/pull/4 we will not have the `TableSource` because the table can be also get data from the external connector.

Reply:
Yes.

The core difference here is that table itself is _a source of truth_ so we must allow people to do anything on it, including synchronize external changes on it or manually update/delete/insert into it.